controllers/clustertopology_controller_compute.go

```diff
@@ -140,9 +141,129 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
+	infra, err := r.getReference(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrap(err, "cluster infrastructure state")
+	}
+	return infra, nil
+}
+
+// getControlPlaneState returns information on the control plane being used by the cluster; If a reference not found,
+// an error is thrown. Also, if the control plane requires infrastructure machines as configured in the clusterClass
+// an error will be thrown if the control plane is found but with no infrastructure machine templates.
+func (r *ClusterTopologyReconciler) getControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getReference(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrap(err, "control plane")
+	}
+	// Some control plane providers may not require infrastructure machines to run. This check returns early if the field
+	// indicating infrastructure is required is not found in the clusterClass of the given cluster
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane reference to infrastructure machine not valid")
+	}
+	infra, err := r.getReference(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrap(err, "control plane infrastructure reference not found")
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries the apiserver for all MachineDeployments linked to a given cluster and managed
+// through a ClusterClass using labels. A cluster may have zero or more MachineDeployments. Zero is expected on first reconcile.
+// If it finds MachineDeployments it inspects them for infrastructure and bootstrap references. Where these are not found
+// the function will throw an error.
+func (r *ClusterTopologyReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *clusterv1.Cluster) ([]machineDeploymentTopologyState, error) {
+	mDeploymentState := make([]machineDeploymentTopologyState, 0)
+
+	md := &clusterv1.MachineDeploymentList{}
+	err := r.Client.List(ctx, md, client.MatchingLabels{
+		clusterv1.ClusterLabelName:         cluster.Name,
+		clusterv1.ClusterTopologyLabelName: "",
+	})
+	if err != nil {
+		return mDeploymentState, errors.Wrap(err, "failed to read machine deployments for managed topology")
+	}
+	for _, m := range md.Items {
+		machineDeployment := m
+		infraRef := &m.Spec.Template.Spec.InfrastructureRef
+		if infraRef == nil {
+			return mDeploymentState, fmt.Errorf("machine deployment %s does not have a reference to a infrastructure machine template", machineDeployment.Name)
```

Comment:
```suggestion
			return mDeploymentState, fmt.Errorf("MachineDeployment %s does not have a reference to a InfrastructureMachineTemplate", machineDeployment.Name)
```
(same/similar for other errors)

Reply:
👍